index.bs

```diff
@@ -1101,31 +1127,116 @@ strong map from handle ids to their corresponding objects.
 OwnershipModel = "root" / "none";
 </pre>
 
-<dfn>RemoteReference</dfn> represents a remote reference to an existing ECMAScript
-object in [=handle object map=] in the given [=Realm=].
+<dfn>RemoteReference</dfn> is either a <code>RemoteObjectReference</code>
+representing a remote reference to an existing ECMAScript object in
+[=handle object map=] in the given [=Realm=], or is a <code>SharedReference</code>
+representing a reference to a [=node=].
 
 <pre class="cddl remote-cddl local-cddl">
 Handle = text;
+SharedId = text;
+</pre>
+
+<pre class="cddl remote-cddl">
+<!-- This is specifically ordered in the order in which matches need to be -->
+<!-- evaluated, since the definitions are overlapping -->
+RemoteReference = (
+  SharedReference //
+  RemoteObjectReference
+)
 
-RemoteReference = {
+SharedReference = {
+   sharedId: SharedId
+   <!-- Ensure that if we have a handle, it at least has the correct type -->
+   ?handle: Handle,
+   Extensible
+}
+
+RemoteObjectReference = {
    handle: Handle,
+   <!-- This shouldn't ever match. The problem is that Extensible would
+   otherwise allow this to match a non-text sharedId -->
+   ?sharedId: SharedId
    Extensible
 }
 </pre>
 
 Issue: handle "stale object reference" case.
 
+Note: if the provided reference has both <code>handle</code> and
+<code>sharedId</code>, the algorithm will ignore <code>handle</code> and respect
+only <code>sharedId</code>.
+
+### Deserialization ### {#data-types-reference-deserialization}
+
 <div algorithm>
-To <dfn>deserialize remote reference</dfn> given |realm| and |remote reference|:
+To <dfn>deserialize remote reference</dfn> given |remote reference|, |realm| and
+|session|:
+
+1. Assert |remote reference| matches the <code>RemoteReference</code> production.
+
+1. If |remote reference| matches the <code>SharedReference</code> production, return
+   [=deserialize shared reference=] given |remote reference|, |realm| and |session|.
 
-1. Let |handle id| be the value of the <code>handle</code> field of |remote reference|.
+1. Return [=deserialize remote object reference=] given |remote reference| and
+   |realm|.
+
+</div>
+
+<div algorithm>
+To <dfn>deserialize remote object reference</dfn> given |remote object reference| and
+|realm|:
+
+1. Let |handle id| be the value of the <code>handle</code> field of
+   |remote object reference|.
 
 1. Let |handle map| be |realm|'s [=handle object map=]
 
 1. If |handle map| does not contain |handle id|, then return [=error=] with
-   [=error code=] [=invalid argument=].
+   [=error code=] [=no such handle=].
+
+1. Return [=success=] with data |handle map|[|handle id|].
+
+</div>
+
+<div algorithm>
+To <dfn>deserialize shared reference</dfn> given |shared reference|, |realm| and
+|session|:
+
+1. Assert |shared reference| matches the <code>SharedReference</code> production.
```

Comment:
We never "need" these asserts. But the idea is to clearly express invariants so that we're less likely to introduce breakage with future edits (we could probably be better at this in general e.g. giving each algorithm parameter an explifit spec type).